Profile driven method layout

davidwrighton opened the pull request:
- Update to mpgo file format to hold weighted call graph data, and exclusive hit sample data
- Implementation of call graph data capture from both jitted and R2R code
- Add mechanism for parsing entrypoints from R2R files into arbitrary type systems insetad of just strings
- Update command line parser in dotnet-pgo to use named properties on class instead of arguments to main method
- Update dotnet-pgo to be able to use a etl.zip file
- Implement infrastructure for performing layout of PE file based on profile guided information
- Implement various profile guided method ordering routines
  - There are a variety of possible algorithms, these are simple to
  implement. More capable ones will generally use the call graph data. As this moves from an experimental feature to a product feature we should remove all but the one shown to be best
  - Sort by exclusive weight
  - Sort by Hot (known to be used) and cold (not known to be used)
  - Sort by Hot (known to be used for more than 128 sample) Warm (known
  to be used) and Cold (not known to be used)
- Since the BulkType logging lock seems to be losing some types, skip the lock for MethodDetails data

davidwrighton committed:
Profile driven layout
- Update to mpgo file format to hold weighted call graph data, and exclusive hit sample data
- Implementation of call graph data capture from both jitted and R2R code
- Add mechanism for parsing entrypoints from R2R files into arbitrary type systems insetad of just strings
- Update command line parser in dotnet-pgo to use named properties on class instead of arguments to main method
- Update dotnet-pgo to be able to use a etl.zip file
- Implement infrastructure for performing layout of PE file based on profile guided information
- Implement various profile guided method ordering routines
  - There are a variety of possible algorithms, these are simple to
  implement. More capable ones will generally use the call graph data.
  - Sort by exclusive weight
  - Sort by Hot (known to be used) and cold (not known to be used)
  - Sort by Hot (known to be used for more than 128 sample) Warm (known
  to be used) and Cold (not known to be used)
- Since the BulkType logging lock seems to be losing some types, skip the lock for MethodDetails data

davidwrighton commented on src/coreclr/src/tools/crossgen2/crossgen2/CommandLineOptions.cs:
```diff
@@ -178,6 +178,14 @@ public static Command RootCommand()
                 {
                     Argument = new Argument<bool>()
                 },
+                new Option(new[] { "--method-layout" }, "Layout file using profile driven optimization assuming the layout algorithm specified. The default value is DefaultSort, which indicates that complex layout is disabled")
```
Need to use string resource for parameter description

MichalStrehovsky commented on src/coreclr/src/tools/Common/Compiler/DependencyAnalysis/SortableDependencyNode.cs:
```diff
@@ -31,6 +31,9 @@ public abstract class SortableDependencyNode : DependencyNodeCore<NodeFactory>,
         public abstract int ClassCode { get; }
 
+        // Custom sort order. Used to override the default sorting mechanics.
+        public int CustomSort = int.MaxValue;
```
Could you move SortableDependencyNode out of Common and into the ReadyToRun specific portion of the compiler?

lindexi replied:
I think it should be a property not a field

davidwrighton replied:
@MichalStrehovsky Could you provide a rationale as to why we should move this from Common? I believe that the CoreRT compiler still uses this exact mechanism, and providing a means to adjust the sort order is likely useful for the CoreRT system as well.

MichalStrehovsky replied:
We don't want externally mutable state on Dependency nodes and this adds mutable state to all of them - I don't want anyone to use that as a precedent. Broken windows theory and all.

CoreRT doesn't need this because the linker is going to shuffle things anyway.

MichalStrehovsky replied:
(If I want to sort something in the middle of the compilation, I better remember that sort order is not deterministic until whatever sets this does it's job. I don't think it's a good property. The sorting should be external to DependencyNodes but I don't want to ask you to do more work, so it's easier to just unshare)

davidwrighton replied:
Ok, fair.

davidwrighton replied:
I've made the SortableDependencyNode into a partial class. Sortable dependency node is tied to quite a lot of the rest of the common code, so it isn't practical to extract.

lindexi commented on src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunSignature.cs:
```diff
@@ -537,6 +554,403 @@ public CorElementType PeekElementType()
+            Options = options;
+        }
+
+        public StringBuilder CurrentBuilder;
```
Should we make this field public?

davidwrighton replied:
We don't actually need this field. I'll delete it.

lindexi commented on src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunSignature.cs:
```diff
@@ -537,6 +554,403 @@ public CorElementType PeekElementType()
+
+            public string GetMethodFromMethodDef(MetadataReader reader, MethodDefinitionHandle handle, string owningTypeOverride)
+            {
+                StringBuilder signaturePrefixBuilder = new StringBuilder();
```
It looks like this is a local variable that is not being used

davidwrighton replied:
Deleted

lindexi commented on src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunSignature.cs:
```diff
@@ -537,6 +554,403 @@ public CorElementType PeekElementType()
+
+            public string GetMethodFromMemberRef(MetadataReader reader, MemberReferenceHandle handle, string owningTypeOverride)
+            {
+                StringBuilder signaturePrefixBuilder = new StringBuilder();
```
It looks like this is a local variable that is not being used

davidwrighton replied:
Deleted

lindexi commented on src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunSignature.cs:
```diff
@@ -537,6 +554,403 @@ public CorElementType PeekElementType()
+                    builder.Append("[INST] ");
+                }
+                builder.Append(method);
+                return method;
```
Why we return the `method` not the `builder`

davidwrighton replied:
That is a bug, thank for the review

trylek commented:
I don't have any strong feeling about InlineSignatureBinary. At some point it seemed to be pretty useful when I and JanV were chasing down the remaining inconsistencies in signature encoding w.r.t. module overrides and such but I doubt anyone has used that feature for a while. After all SignatureBinary should still work unless I'm mistaken. I wouldn't be opposed to throwing ISB away altogether if it's causing you trouble. For lowercase primitive types, thanks for doing that, that sounds great to me.

trylek commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunFileLayoutOptimizer.cs:
```diff
@@ -0,0 +1,187 @@
+            if (_methodLayoutAlgorithm == ReadyToRunMethodLayoutAlgorithm.DefaultSort)
+                return nodes;
+
+            Func<List<MethodWithGCInfo>> sortedMethods = null;
```
Non-blocking nit - considering we use sortedMethods only once after the if statement to populate a list, is there any reason why using these auxiliary lambdas is more elegant than directly defining a List<MethodWithGCInfo> at this point and filling it with elements in the individual conditional blocks?

davidwrighton committed:
Fix review comments

cshung commented on src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunSignature.cs:
```diff
@@ -364,15 +366,25 @@ private string EmitString(StringHandle handle)
 
+    public interface IR2RSignatureTypeProvider<TType, TMethod, TGenericContext> : ISignatureTypeProvider<TType, TGenericContext>
+    {
+        public TType GetCanonType();
```
It looks like we don't need to declare the method as public here.
This won't compile if we set the language version to be 7.3

davidwrighton replied:
:) I've been accidentally adding public to interface methods since C# 1.0. As of the latest compilers it finally just lets me do it. I find it satisfying. (I will admit, when I was in the language design meeting where we added the feature I was kinda happy, even though it means that I will accidentally violate the coding standard somewhat often.)

However, as you say, it certainly isn't needed. I'll remove them.

cshung commented on src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunSignature.cs:
```diff
@@ -612,41 +1026,41 @@ private void EmitInlineSignatureBinaryBytes(StringBuilder builder, int count)
             return value;
         }
 
         private int ReadIntAndEmitInlineSignatureBinary(StringBuilder builder)
```
It appears that this method is no longer called by anyone, we can delete it.

cshung commented on src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunSignature.cs:
```diff
@@ -612,41 +1026,41 @@ private void EmitInlineSignatureBinaryBytes(StringBuilder builder, int count)
             return value;
         }
 
         private void EmitSignatureBinaryFrom(StringBuilder builder, int startOffset)
```
Should this be `EmitSignatureBinaryForm`?

davidwrighton replied:
Eh, I actually think the From naming is more accurate. It specifies to dump a signature from and offset to the current offset. In any case, I haven't changed those names in this PR, and given that the change is already too large, I don't want to make unnecessary changes.

trylek reviewed (approved):
To me this looks good, awesome work, thank you!

cshung commented on src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs:
```diff
@@ -449,12 +449,46 @@ private void EnsureMethods()
+            }
+        }
+
+        public Dictionary<TMethod, ReadyToRunMethod> GetCustomMethodToRuntimeFunctionMapping<TType, TMethod, TGenericContext>(IR2RSignatureTypeProvider<TType, TMethod, TGenericContext> provider)
```
Can we make this an `IReadOnlyDictionary` to keep this public interface immutable?
We should not allow changing the dictionary anyway.

davidwrighton replied:
Specifically, this Dictionary isn't cached anywhere, and actually *is* safely mutable, but keeping the public api as providing an immutable surface isn't unreasonable. I'll make that change.

davidwrighton committed:
Address feedback

davidwrighton committed:
Move the custom sort behavior out to crossgen2 specific code
- Use a partial class as SortableDependencyNode is fairly intertwined with the common codebase

davidwrighton committed:
Specify command line option string via string resource
